### PR TITLE
Fix bucketing of strings with non-BMP characters

### PR DESCRIPTION
**Pull request by findepi**
Fixes https://github.com/apache/iceberg/issues/2837

### CODE REVIEW

**Comment by findepi** [comment]
cc @alexjo2144 @losipiuk @joshthoward @electrum

File: api/src/test/java/org/apache/iceberg/transforms/TestBucketing.java [hunk 1]
```diff
@@ -215,6 +215,24 @@ public void testString() {
...
+    Bucket<CharSequence> bucketFunc = Bucket.get(Types.StringType.get(), 100);
+
+    Assert.assertEquals("String hash should match hash of UTF-8 bytes",
+        hashBytes(asBytes), bucketFunc.hash(string));
+
+    Assert.assertNotEquals("It looks like Guava has been updated and now contains a fix for " +
+                    "https://github.com/google/guava/issues/5648. Please resolve the TODO in BucketString.hash " +
+                    "and remove this assertion",
+            hashBytes(asBytes),
+            MURMUR3.hashString(string, StandardCharsets.UTF_8).asInt());
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/transforms/TestBucketing.java:233]: Nit: Looks like indentation is off for this statement.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good to me. I agree with the choice to detect cases where `hashString` can't be used since converting the string to bytes will allocate a new `ByteBuffer` and copy into it.

**Comment by rdblue** [comment]
@findepi, looks like I don't have permission to update the indentation in the test. Could you fix that and I'll merge this?

File: api/src/test/java/org/apache/iceberg/transforms/TestBucketing.java (hunk 1 above)
Comment (findepi) [reply api/src/test/java/org/apache/iceberg/transforms/TestBucketing.java:233]: eventually my IDE is not against me on this :) 

fixed

**Review by rdblue — APPROVED** [review]

**Comment by findepi** [comment]
Thanks @rdblue

**Comment by rdblue** [comment]
Thanks for fixing this, @findepi! Nice work.

**Pull request merged**